File: src/haskell/2018/Day1.hs
{-|
Module      : Day1
Description : Day 1: Chronal Calibration

<https://adventofcode.com/2018/day/1>
-}

module Day1 where

import Advent
import Data.Set (Set, empty, member, insert)

parseNum :: String -> Int
parseNum ('+':r) = read r
parseNum r = read r

firstRepeated :: Set Int -> [Int] -> Maybe Int
firstRepeated seen (x:xs)
    | member x seen = Just x
    | otherwise = firstRepeated (insert x seen) xs
firstRepeated _ _ = Nothing

main = do
    input <- parsedInput (2018, 1) (map parseNum . lines)
    print $ sum input
    printMaybe $ firstRepeated empty $ scanl (+) 0 $ cycle input

File: src/python/2018/day1.py
# Advent of Code 2018, day 1: Chronal Calibration

import itertools

from functools import reduce
from helpers import int_readlines

data = int_readlines(2018, 1)

def puzzle2():
    seen = set([0])
    for i in itertools.accumulate(itertools.cycle(data)):
        if i in seen:
            return i
        seen.add(i)

print(reduce((lambda x,y: x+y), data))
print(puzzle2())
